src/main/java/com/google/devtools/build/lib/analysis/AnalysisResult.java

```diff
@@ -123,6 +126,14 @@ public PackageRoots getPackageRoots() {
     return targetsToSkip;
   }
 
+  /**
+   * Returns the incompatible targets that were explicitly requested on the command line.
+   */
+  // TODO(phil): Document this properly.
```

Comment:
+1. :)

I'd especially like to be clear on this because it's also possible that some of `targetsToBuild` will fail (because they just don't build right). So on first glance it reads strangely to see that there's some set of targets we somehow already know we want to fail (which begs the question "why even try?") and that for some reasons these go in a separate place then the `targetsToBuild` that also fail.

I do see the logic here. Ideally we could sort this out before we even get to AnalysisResult. But I think you've clarified we can't?

Reply:
Yeah, at the time I couldn't figure out a way to sort this out before getting to AnalysisResult. At least not while trying to keep the changes small.

I'll spend some more time on this to see if I missed something. Basically I couldn't figure out a way for bazel to skip the targets _and_ mark them as failed at the same time.

Reply:
Actually, I have another idea. I think it could simplify the entire feature if we apply it carefully enough.

Bazel already has logic that lets a rule look at itself and determine if it should succeed or fail. This is the build's analysis phase, and it's the standard code path everything in `targetsToBuild` goes through. Rather than work around that structure, why don't we work *within* it? i.e. just add the concept of `targetsToFail` as just another piece of code in the existing analysis logic, which already has the job of seeing if targets should succeed or fail?

We'd have to patch together the logic in a different set of files. But I think we can isolate pretty closely what needs to change so we don't get caught up in the intimating complexity of `ConfiguredTargetFunction` (which is a HUGE file that basically directs the analysis phase for a target and does a million things to accomplish that).

If you look in `ConfiguredTargetFunction`, you see it has this call:

https://github.com/bazelbuild/bazel/blob/5fa85c8e9e32d9328be356e76e51d739e14fd0da/src/main/java/com/google/devtools/build/lib/skyframe/ConfiguredTargetFunction.java#L333

If you look at that file you see it has this purpose:

https://github.com/bazelbuild/bazel/blob/b100b1d65f45864efb8e9855cdce6e3de2deec66/src/main/java/com/google/devtools/build/lib/analysis/ResolvedToolchainContext.java#L41-L43

which is a pretty similar purpose to what you're doing.

What if next to the `ResolvedToolchainContext.load(` call we have a corresponding call to `TopLevelConstraintSemantics.checkExplicitlySpecifiedTarget` (or whatever name) that does the same platform compatibility check and throws an Exception if the target is incompatible? I think this would be a lot simpler than what's currently happening in `TopLevelConstraintSemantics`, and we can also directly pass in the environment info necessary to get the toolchain context, without having to jump through hoops as `TopLevelConstraintSemantics` does now. 

If it throws some sort of `IncompatiblePlatformException`, we just need to process that like at https://github.com/bazelbuild/bazel/blob/5fa85c8e9e32d9328be356e76e51d739e14fd0da/src/main/java/com/google/devtools/build/lib/skyframe/ConfiguredTargetFunction.java#L409-L415.

which just repackages it as a `ConfiguredValueCreationException`. Or we can directly throw a `ConfiguredValueCreationException`. The main point is that once that exception is thrown, the target will automatically fail with a clear message, of your choosing, and I think everything just works.

This has the added benefit (?) of automatically handling non-top-level targets, with no extra work. Did you want to avoid doing that now because you thought it would create problems, or because of the presumed extra work involved? If the latter, this approach would be a great win.

I realize I'm throwing a ton of stuff at you in this comment. Just let me know your basic thoughts on it, clarification questions, or if you see any holes. Happy to iterate in more detail since I know it covers a bunch of complex logic.

Reply:
Ha! And here I thought I was getting closer :stuck_out_tongue: 

That _sounds_ really awesome. Let me stare at it for a while to see if I can understand it.

> Did you want to avoid doing that now because you thought it would create problems, or because of the presumed extra work involved?

Mostly because we figured it'd be too much work for a single patch. However, what you're saying intrigues me a lot. I'll definitely look into it.

Reply:
Sounds good. Please, please feel free to ask me for more clarification if anything isn't obvious. My goal isn't to discourage you, and I know comments like "let's refactor everything in what's already a complicated, hard-fought change" risks that.  

I really see a cleaner and simple solution all around if this approach works. With luck, most of the refactoring will be removing the complicated extra boilerplate you had to add in the last iteration (like the graph access stuff). As long as we're locked into a good approach I'll do everything I can to help carry it through.

Reply:
Definitely not feeling discouraged. This is exciting :smile: 
I'll definitely ask you for clarification as needed. I suspect it'll take me a few days to wrap my head around it, that's all.

Reply:
Looks like making `ResolvedToolchainContext` depend on `TopLevelConstraintSemantics` creates a dependency loop because the `analysis_cluster` target is a *huge* collection of files.

I'll first have to spend some time to break that apart into smaller libraries.

Reply:
Here's another interesting entry point for the new approach:

If we did the checking at the line before https://github.com/bazelbuild/bazel/blob/83221e3be280c4f9ddc88d6e8e5e3824dd5a5db8/src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTargetFactory.java#L445

we could pass the checking code a `RuleContext`. `RuleContext` has this method:

https://github.com/bazelbuild/bazel/blob/5037537571abe1c9be7ff552e0dbd09381f0fb92/src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java#L1248

which looks like a pretty straightforward call to see if the target platform has a constraint setting or not. `RuleContext` also has access to the underlying `Rule` and its attributes, so it's easy to look at `target_compatible_with` from there too.

If the check fails, I think all it has to do is create an `AnalysisFailure` object and Bazel automatically handles the rest.

Reply:
Maybe that could be simple enough to break up some of the complexities you're finding with `TopLevleConstraintSemantics`?

Reply:
Haha, let me look at that. I've been untangling the `analysis_cluster` library and man, it's one deep rabbit hole.
I'll pause on that and look into the alternative entry point here. Thanks!

Reply:
Okay, this is really neat. Thanks!
Accessing the `target_compatible_with` attribute is pretty straight forward.

Admittedly, I'm struggling to convert the `List<Label>` from the attribute into a list of `ConstraintValueInfo` instances that I can use with `targetPlatformHasConstraint`. If you have any hints, I'll be glad to take them :)

Otherwise I'm sure I'll figure it out sooner or later. We did something similar with the first approach already, but my skills at converting between the various classes inside Bazel is still subpar.

This is super exciting :smile:

Reply:
I was also a bit sure about converting to `ConstraintValueInfo`. :) But I *think* there's a reasonable way. CC'ing @katre for domain expert perspective on that but I'll also spelunk into some code to see if I can find anything...

Reply:
From the depending rule, can you call

    ruleContext.getPrerequisites("target_compatible_with", TransitionMode.TARGET)

then for each result (each one should correspond to one of the constraint values) call

    result.getProvider(ConstraintValueInfo.class)

?

With that approach, rather than trying to parse labels into providers, you can use the standard APIs of getting a target's deps (the actual deps the labels refer to) and extracting the provider directly from them.

Even more direct, check out https://github.com/bazelbuild/bazel/blob/b44ea8cd670e753635ca6be209a4ef9153422b76/src/main/java/com/google/devtools/build/lib/rules/config/ConfigSetting.java#L169-L177

which uses some helper methods in `PlatformProviderUtils` to automate even further.

Reply:
A couple of comments:

- I broke up everything in `build_base`, so I can fairly confidently state that `analysis_cluster` is going to be very difficult to detangle. Code changes will be required, it will be a mess. It might be possible to extract one or two classes from it, but the core code (things like `ConfiguredTargetFactory`, everything under `configuredtargets`, the core `config` classes) aren't going to be splitable. Can you clarify the problem you're seeing using `TopLevelConstraintSemantics`? One option is to merge that class back into `analysis_cluster` if needed.
- Using `getDirectPrerequisites` and `PlatformProviderUtils` is the easiest way to get instances of `ConstraintValueInfo`. See how the calls in [`Platform`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/platform/Platform.java;drc=8b856f5484f0117b2aebc302f849c2a15f273310;l=137) work for an example.

This is exciting work! I'm happy to see it coming along.

Reply:
I wonder how much of `TopLevelConstraintSemantics` we even need with approach laid out in these last few comments. Maybe it really just comes down to using `PlatformProviderUtils` get to get the `ConstraintInfo`, checking how that matches the current platform, creating a `ruleContext.AttributeError` or `AnalysisFailure`, and that's it?

It'd be really exciting if there's an opportunity to heavily reduce the complexity of `TopLevelConstraintSemantics`.

Reply:
I think I'm missing something fundamental.

I have the following patch:
```patch
@@ -442,6 +447,21 @@ public final class ConfiguredTargetFactory {
                     prerequisiteMap.values()))
             .build();

+    if (!"toolchain".equals(rule.getRuleClass())) {
+      if (ruleContext.attributes().has("target_compatible_with")) {
+        Iterable<ConstraintValueInfo> constraintValues =
+            PlatformProviderUtils.constraintValues(
+                ruleContext.getPrerequisites(
+                    "target_compatible_with", TransitionMode.DONT_CHECK));
+        for (ConstraintValueInfo cvi : constraintValues) {
+          if (!ruleContext.targetPlatformHasConstraint(cvi)) {
+            ruleContext.attributeError(
+                "target_compatible_with", "Rule is not compatible with " + cvi.label().toString());
+          }
+        }
+      }
+    }
+
     List<NestedSet<AnalysisFailure>> analysisFailures = depAnalysisFailures(ruleContext);
     if (!analysisFailures.isEmpty()) {
       return erroredConfiguredTargetWithFailures(ruleContext, analysisFailures);
```
which I _think_ does what you suggested. The problem that I'm seeing is that there's no way to differentiate this from a regular analysis failure (i.e. build not completing successfully) and an analysis "failure" that should cause the target to be skipped.

In other words, I don't see a way to make bazel skip the targets that fail the check. It'll stop the build due to the analysis failure.

Maybe we could add some skip/no-skip info in `RuleContext` and extract that in `TopLevelConstraintSemantics` somehow. That's my current line of thinking at least. Will resume on the weekend on this.

Reply:
That being said, really appreciate the hints from you guys @gregestren and @katre !

Reply:
Excellent point.

I had to think about that for a while. But I think I have a good response (if this is different than what you were exploring over the weekend, apologies).

There are two primary principles I'm applying to my thinking:

1) It's great to leverage the existing analysis framework as much as possible. The more we can do that, the more general, flexible, and supportable the result will be, with the least amount of new code complexity.

2) Targets configured in the existing framework have no idea whether they're top-level or not, or explicitly specified at the command line vs. part of `//:...`.

Principle 1 tells me it's not the target's job to figure out if it's a skip or failure, because principle 2 says it doesn't have access to that knowledge.

But the target *does* know if it's compatible or not (i.e. the code snippet in your above comment). While it doesn't know what to do with that information, it does know how to report that information for other code to decide. The standard way by which analyzed targets communicate information about themselves is through *providers*.

So:

1) Replace the `ruleContext.attributeError` with returning a new `IncompatiblePlatformProvider` class (that can be at trivial class that implements `TransitiveInfoProvider`. Here's an example that creates a trivial configured target that sets a provider:

https://github.com/bazelbuild/bazel/blob/7a3eb4f61c65e9ba4a8f9be68867f40dc5166845/src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTargetFactory.java#L730-L739

2) Go back to the code that *does* know what's top-level explicit and replace:

https://github.com/bazelbuild/bazel/blob/7a3eb4f61c65e9ba4a8f9be68867f40dc5166845/src/main/java/com/google/devtools/build/lib/analysis/BuildView.java#L426-L431

with a loop that goes through `skyframeAnalysisResult.getConfiguredTargets` and adds each configured target where `ct.getProvider(IncompatiblePlatformProvider.class) != null` to `targetsToSkip` if it should be skipped, else triggers an error.

So this would basically chop what `TopLevelConstraintSemantics` does today into two pieces. The piece that figures out if a target is compatible or not goes into the standard rule analysis framework. The piece that figures out if that's a skip or error happens at the same place `TopLevelConstraintSemantics` is called today, where instead of computing all the results itself it just looks at the results already computed from the first part.
That's sort of similar to what `TopLevleConstraintSemantics` does internally today:

https://github.com/bazelbuild/bazel/blob/802551e0c264fa2287f3c6f9cd8753e48c287c47/src/main/java/com/google/devtools/build/lib/analysis/constraints/TopLevelConstraintSemantics.java#L155-L156

The only caveat I'd add is support for the `--keep_going` flag, which means don't fail the whole build if just one target errors. When that flag is on Bazel logs an error event instead of an exception:

https://github.com/bazelbuild/bazel/blob/a111a2150abe775a001ce0ee756d1bd7f8677bf7/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeBuildView.java#L480-L494

Reply:
I also like that this approach explicitly separates *whether* a target is compatible with *what* to do about it.

That creates all kinds of interesting opportunities to play with future behavior however we like.

For example, if we want a top-level target to succeed when its dependency target is incompatible, we just have it ignore the `IncompatiblePlatformProvider` its dep reports. But if we want the top-level target to fail too, we have it propagate the provider up.

Reply:
That's really neat! I very much like the provider approach. I was working with lots of casts and it wasn't pretty.

Let me give that a try. It looks really promising :smile:

Reply:
I could use some help understanding how to work with providers. It's pretty easy to do the following in case of an incompatiblity:
```java
RuleConfiguredTargetBuilder builder = new RuleConfiguredTargetBuilder(ruleContext);
builder.add(IncompatiblePlatformProvider.class, IncompatiblePlatformProvider.simple());
builder.add(RunfilesProvider.class, RunfilesProvider.simple(Runfiles.EMPTY));
return builder.build();
```
However, that generates runtime errors like:
```
java.lang.IllegalStateException: Test rules must have runfiles
```
So I went down the route of trying to add the provider to the `ConfiguredTarget` instances that get generated in `ConfiguredTargetFactory.createRule()`.
It's relatively straight forward (though ugly) to inject an additional provider for the starlark code path here:
https://github.com/bazelbuild/bazel/blob/7a3eb4f61c65e9ba4a8f9be68867f40dc5166845/src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTargetFactory.java#L469-L480
However, for the built-in rules code path (e.g. `ShBinary.create()`) here, it seems a lot trickier (or at least a lot more invasive) to attempt injecting a provider:
https://github.com/bazelbuild/bazel/blob/7a3eb4f61c65e9ba4a8f9be68867f40dc5166845/src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTargetFactory.java#L481-L489
Then I tried to see if I can attach a provider to an existing `ConfiguredTarget`, but I concluded that that goes against the spirit of what a configured target is. (I.e. it's not really configured if I want to attach more providers to it).

I'm wondering how important it is for the returned `ConfiguredTarget` to be realistic. I.e. does the `ConfiguredTarget` instance affect `bazel cquery` results for example?

Anyway, that's where I'm at right now. If you happen to have some pointers, I'd appreciate them :)  Haven't decided what angle to approach this from next.

Reply:
It's a good point about `blaze cquery`. But for now I think it's better for it not to be realistic, i.e. to stick with the top approach of instantiating and returning your own builder. 

The reason is that it'll save some processing time: analyzing the full configured target takes some amount of time and if we already know it's going to fail because it's incompatible that's just wasted work.

It's always something we can change in another PR if a clear need for the "accurate" version ever came out.

Regarding the `Test rules must have runfiles` error, can you "fake it out" by calling `builder.setRunfilesSupport` with `RunfilesSupport.withExecutable` with maybe a `Runfiles.EMPTY` and `ruleContext.createOutputArtifact()`? 

A quick glance through the code suggests that runfiles check (on tests only) is the only extra check the system does?

Reply:
Thanks for the clarification, not worrying about realistic runfiles sounds good to me :)
I'll pursue that path, thanks a lot!

Reply:
The provider approach works really well. Thanks a lot for all the pointers. I really appreciate it. The patch is looking quite simple now.

I've got most of the functionality restored. The one missing part is where explicitly requested targets on the command line should be marked as failed instead of skipped. I don't see a simple way to do that other than what's in this review already (i.e. via `targetsToFail`).

I'll spend some more time to think if there's a better way. If I cannot, then I'll aim to have another patch uploaded to the review here before next week.

Reply:
Happy to hear overall.

Re: explicit vs. not: I look forward to seeing the next patch as a whole, and happy to offer more input from that vantage point. I acknowledge that's a legitimate concern we still need to properly resolve. I have some ideas in my head, but i'll hold off until the patch is ready. :)